File: src/main/java/com/ada/conversation/application/SelectModelUseCase.java
package com.ada.conversation.application;

import com.ada.conversation.application.dto.*;
import com.ada.conversation.application.port.in.RequestFilter;
import com.ada.model.application.port.out.ModelSelectionStrategy;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class SelectModelUseCase {
  private final List<RequestFilter> filters;
  private final List<ModelSelectionStrategy> strategies;

  public ModelSelection execute(ChatRequest r) {
    var current = r;
    for (var f : filters) if (f.supports(current)) current = f.apply(current);
    for (var s : strategies) if (s.supports(current)) return s.select(current);
    throw new IllegalStateException("No model selection strategy supports the request");
  }
}

File: src/main/kotlin/com/ada/conversation/application/SelectModelUseCase.kt
package com.ada.conversation.application

import com.ada.conversation.application.port.`in`.ChatRequest
import com.ada.conversation.application.port.`in`.RequestFilter
import com.ada.model.application.port.out.ModelSelection
import com.ada.model.application.port.out.ModelSelectionStrategy
import org.springframework.stereotype.Service

@Service
class SelectModelUseCase(
    private val filters: List<RequestFilter>,
    private val strategies: List<ModelSelectionStrategy>,
) {
    fun execute(request: ChatRequest): ModelSelection {
        val filteredRequest = filters
            .filter { it.supports(request) }
            .fold(request) { current, filter -> filter.apply(current) }

        return strategies
            .firstOrNull { it.supports(filteredRequest) }
            ?.select(filteredRequest)
            ?: error("No model selection strategy supports the request")
    }
}
